planner/core/plan_stats.go

```diff
@@ -124,24 +144,103 @@ func SyncWaitStatsLoad(plan LogicalPlan) error {
+// virtual columns we want will be there.
+// However, in some cases (the mv index case now), the expressions are not replaced with the virtual columns before here.
+// Instead, we match the expression in the query against the expression behind the virtual columns after here when
+// building the access paths. This means we are unable to known what virtual columns will be needed by just extracting
+// the Column from the expressions here. So we need to manually collect the virtual columns that may be needed.
+//
+// Note 1: As long as a virtual column depends on the needed columns, it will be collected. This could collect some virtual
+// columns that are not actually needed.
+// It's OK because that's how sync loading is expected. Sync loading only needs to ensure all actually needed stats are
+// triggered to be loaded. Other logic of sync loading also works like this.
+// If we want to collect only the virtual columns that are actually needed, we need to make the checking logic here exactly
+// the same as the logic for generating the access paths, which will make the logic here very complicated.
+//
+// Note 2: Only direct dependencies are considered here.
+// If a virtual column depends on another virtual column, and the
+// latter depends on the needed columns, then the former will not be collected.
+// For example: create table t(a int, b int, c int as (a+b), d int as (c+1)); If a is needed, then c will be collected,
+// but d will not be collected.
+// It's because currently it's impossible that statistics related to indirectly depending columns are actually needed.
+// If we need to check indirect dependency some day, we can easily extend the logic here.
+func CollectDependingVirtualCols(tblID2Tbl map[int64]table.Table, neededItems []model.TableItemID) []model.TableItemID {
+	generatedCols := make([]model.TableItemID, 0)
+
+	// group the neededItems by table id
+	tblID2neededCols := make(map[int64][]model.TableItemID, len(tblID2Tbl))
+	for _, item := range neededItems {
+		if item.IsIndex {
+			continue
+		}
+		tblID2neededCols[item.TableID] = append(tblID2neededCols[item.TableID], item)
+	}
+
+	// process them by table id
+	for tblID, colItems := range tblID2neededCols {
+		tbl := tblID2Tbl[tblID]
+		if tbl == nil {
+			continue
+		}
+		// collect the needed columns on this table into a set for faster lookup
+		colNameSet := make(map[string]struct{}, len(colItems))
+		for _, colItem := range colItems {
+			name := tbl.Meta().FindColumnNameByID(colItem.ID)
+			if name == "" {
+				continue
+			}
+			colNameSet[name] = struct{}{}
+		}
+		// iterate columns in this table, and collect the virtual columns that depend on the needed columns
+		for _, col := range tbl.Cols() {
+			// only handles virtual columns
+			if !col.IsGenerated() || col.GeneratedStored {
+				continue
+			}
+			// If this column is already needed, then skip it.
+			if _, ok := colNameSet[col.Name.L]; ok {
```

Comment:
Can we use the ID of the col to as the map key here? IntMap is always better than StringMap